pkg/chunk/purger/tombstones.go

```diff
@@ -0,0 +1,293 @@
+package purger
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const tombstonesReloadDuration = 15 * time.Minute
+
+// TombstonesSet holds all the pending delete requests for a user
+type TombstonesSet struct {
+	tombstones                               []DeleteRequest
+	oldestTombstoneStart, newestTombstoneEnd model.Time // Used as optimization to find whether we want to iterate over tombstones or not
+}
+
+// TombstonesLoader loads delete requests and gen numbers from store and keeps checking for updates.
+// It keeps checking for changes in gen numbers, which also means changes in delete requests and reloads specific users delete requests.
+type TombstonesLoader struct {
+	tombstones    map[string]*TombstonesSet
+	tombstonesMtx sync.RWMutex
+
+	deleteStore *DeleteStore
+	quit        chan struct{}
+}
+
+// NewTombstonesLoader creates a TombstonesLoader
+func NewTombstonesLoader(deleteStore *DeleteStore) *TombstonesLoader {
+	tl := TombstonesLoader{
+		tombstones:  map[string]*TombstonesSet{},
+		deleteStore: deleteStore,
+	}
+	go tl.loop()
+
+	return &tl
+}
+
+// Stop stops TombstonesLoader
+func (tl *TombstonesLoader) Stop() {
+	close(tl.quit)
+}
+
+func (tl *TombstonesLoader) loop() {
+	tombstonesReloadTimer := time.NewTicker(tombstonesReloadDuration)
+	for {
+		select {
+		case <-tombstonesReloadTimer.C:
+			err := tl.reloadTombstones()
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error reloading tombstones", "err", err)
+			}
+		case <-tl.quit:
+			return
+		}
+	}
+}
+
+func (tl *TombstonesLoader) reloadTombstones() error {
+	// check for updates in loaded gen numbers
+	tl.tombstonesMtx.Lock()
+
+	userIDs := make([]string, 0, len(tl.tombstones))
```

Comment:
From what I can see, we're only reloading or loading tombstones for userIDs that are already there. How are we getting new userIDs in, in the first place?

Reply:
Consumers always call [GetPendingTombstones](https://github.com/cortexproject/cortex/pull/2214/files/4b75b8e685a90acdf52b7ace02e91ac288bfbbd5#diff-c0f1a864b800713f20d644cfca5d25c0R89) which checks whether it has loaded delete requests already, if not it calls [loadPendingTombstones](https://github.com/cortexproject/cortex/pull/2214/files/4b75b8e685a90acdf52b7ace02e91ac288bfbbd5#diff-c0f1a864b800713f20d644cfca5d25c0R99) which loads it.
Does it make sense?